File: java/leetcode-explore/algorithmInterview/dp/BinaryTreeMaxPathSum.java
package algorithmInterview.dp;

import java.util.Arrays;
import java.util.Collections;

public class BinaryTreeMaxPathSum {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    private int res = Integer.MIN_VALUE;

    public int recursion(TreeNode root) {
        if (root == null) return 0;
        int left = recursion(root.left);
        int right = recursion(root.right);
        res = Math.max(res, Collections.max(Arrays.asList(root.val + left + right, root.val + left, root.val + right, root.val)));
        return Math.max(root.val, Math.max(left, right) + root.val);
    }

    public int maxPathSum(TreeNode root) {
        if (root == null) return 0;
        res = Integer.MIN_VALUE;
        recursion(root);
        return res;
    }
}

File: go/tencent/sort_and_search/binary_tree_max_path_sum.go
package main

import (
	"fmt"
	"math"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func main() {
	root := new(TreeNode)
	root.Val = -3
	fmt.Println(maxPathSum(root))
}

var res int = math.MinInt64

func recursionMPS(root *TreeNode) int {
	if root == nil {
		return 0
	}

	left, right := recursionMPS(root.Left), recursionMPS(root.Right)
	res = max(res, left+root.Val, right+root.Val, left+right+root.Val, root.Val)
	return max(max(left, right)+root.Val, root.Val)
}

func maxPathSum(root *TreeNode) int {
	res = math.MinInt64
	recursionMPS(root)
	return res
}

func max(nums ...int) int {
	if nums == nil || len(nums) <= 0 {
		return -1
	}
	res := math.MinInt64
	for _, num := range nums {
		if num > res {
			res = num
		}
	}
	return res
}
